File: packages/ui/src/tree-guides.ts
// EXP-965: the TREE CONNECTOR rule — every nested list on every client draws
// Linear-style guide lines instead of bare indentation. ONE pure function,
// mirrored ×4 (desktop `domain::tree_guides`, iOS `TreeGuides.swift`, Android
// `TreeGuides.kt`) so a session tree, a PR stack and a run tree all connect
// the same way.
//
// The geometry (shared):
//
//   * indent stays 14px per level; a row at depth `d` is padded to
//     `12 + 14 * d`;
//   * gutter level `L` is the 14px band from `12 + 14 * L` to `12 + 14*(L+1)`,
//     its centre (`12 + 14*L + 7`) aligned with the horizontal centre of the
//     leading glyph box of a row AT that depth — so a child's elbow hangs off
//     its parent's glyph;
//   * a row at depth `d ≥ 1` draws the ELBOW in gutter level `d - 1`: a 1px
//     vertical from the row's top edge down to its vertical centre, a 5px
//     rounded corner turning right, and a 1px stub out to the gutter's right
//     edge (just before the child's own glyph);
//   * that vertical continues to the row's BOTTOM edge (a tee) when the row is
//     not the LAST child of its parent among the visible rows;
//   * for every ancestor level whose subtree continues after this row, a
//     straight full-height vertical at that level's gutter centre;
//   * a parent row draws nothing of its own; a folded subtree draws nothing.

/** Guide geometry for ONE row, in gutter LEVELS (not pixels). */
export interface TreeGuide {
  /** The gutter level the elbow sits in (`depth - 1`); `null` for a root. */
  elbowAt: number | null
  /** The elbow's vertical continues past the centre: a later sibling follows. */
  tee: boolean
  /** Ancestor gutter levels carrying a straight full-height line, ascending. */
  passThrough: number[]
}

/** 14px per nesting level — the ×4 indent. */
export const TREE_INDENT = 14
/** The first gutter starts here: `ListRow`'s own 12px left padding. */
export const TREE_BASE = 12
/** The elbow's corner radius. */
export const TREE_RADIUS = 5

/** The x centre of gutter level `level`. `base` is the list's own left
 * padding — `TREE_BASE` for every list row, `0` for the few nestings that
 * indent from a container's edge (the PR stack inside the graph overlay). */
export function treeGuideCentre(level: number, base = TREE_BASE): number {
  return base + TREE_INDENT * level + TREE_INDENT / 2
}

/**
 * Whether the subtree hanging off gutter `level` continues AFTER row `index`
 * — i.e. another row at depth `level + 1` follows before the walk leaves that
 * subtree (a row at depth `≤ level`).
 */
function continuesAfter(
  depths: readonly number[],
  index: number,
  level: number
): boolean {
  for (let at = index + 1; at < depths.length; at += 1) {
    const depth = depths[at]!
    if (depth <= level) return false
    if (depth === level + 1) return true
  }
  return false
}

/**
 * The guides for a flat list of VISIBLE rows, given their depths (a folded
 * subtree is simply absent from the array, which is what makes a collapsed
 * parent draw nothing below it).
 */
export function treeGuides(depths: readonly number[]): TreeGuide[] {
  return depths.map((depth, index) => {
    if (depth <= 0) return { elbowAt: null, tee: false, passThrough: [] }
    const passThrough: number[] = []
    for (let level = 0; level <= depth - 2; level += 1) {
      if (continuesAfter(depths, index, level)) passThrough.push(level)
    }
    return {
      elbowAt: depth - 1,
      tee: continuesAfter(depths, index, depth - 1),
      passThrough,
    }
  })
}

/** Nothing to draw — a root row with no ancestors still carrying a line. */
export function treeGuideIsEmpty(guide: TreeGuide | null | undefined): boolean {
  return (
    !guide || (guide.elbowAt === null && guide.passThrough.length === 0)
  )
}

File: apps/desktop/crates/domain/src/tree_guides.rs
//! EXP-965 — the TREE CONNECTOR every nested list draws instead of bare
//! indentation (web `packages/ui/src/tree-guides.ts`, iOS
//! `ExpCore/Sources/Domain/TreeGuides.swift`, Android `domain/TreeGuides.kt`
//! — the same rule and the same tests).
//!
//! A nested list is a sequence of VISIBLE rows with a depth each (whatever a
//! fold left standing). From that sequence alone the connector falls out:
//!
//! * a row at depth `d ≥ 1` draws an ELBOW in gutter column `d - 1` — the
//!   band its PARENT's leading glyph sits in — running from the row's top
//!   edge down to its vertical centre and turning right into a short stub;
//! * the elbow becomes a TEE when the row is not its parent's last visible
//!   child: the vertical carries on to the row's bottom edge;
//! * every ANCESTOR level whose subtree continues below this row draws a
//!   straight PASS-THROUGH line for the full row height.
//!
//! Pure, so the geometry (which is per-platform) never has to re-derive the
//! facts — and so the rule is one unit test rather than four eyeballings.

/// The connector one row draws, in GUTTER LEVELS (a level is one indent step;
/// the caller turns a level into an x with its own padding and pitch).
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Guides {
    /// The level whose gutter carries this row's elbow — `depth - 1`, and
    /// `None` for a root row (depth 0 draws nothing at all).
    pub elbow_at: Option<usize>,
    /// The elbow's vertical runs the WHOLE row height: another sibling
    /// follows below.
    pub tee: bool,
    /// Ancestor levels (all `< depth - 1`) that draw a full-height line
    /// because that ancestor has a later child.
    pub pass_through: Vec<usize>,
}

impl Guides {
    /// The row's nesting depth — the indent its list pads by. Derived, so a
    /// row carries its connector INSTEAD of a loose depth rather than beside
    /// one (the two can then never disagree).
    pub fn depth(&self) -> usize {
        self.elbow_at.map_or(0, |level| level + 1)
    }
}

/// The connector for every row of a visible, tree-ordered `depths` sequence.
///
/// One backward pass: `seen[x]` means "a later row at depth `x` exists with no
/// shallower row in between", which is exactly "that ancestor has another
/// child below". A row at depth `d` then closes every deeper level.
pub fn guides_for(depths: &[usize]) -> Vec<Guides> {
    let max_depth = depths.iter().copied().max().unwrap_or(0);
    let mut seen = vec![false; max_depth + 2];
    let mut out = vec![Guides::default(); depths.len()];
    for (index, &depth) in depths.iter().enumerate().rev() {
        if depth > 0 {
            out[index] = Guides {
                elbow_at: Some(depth - 1),
                tee: seen[depth],
                // Levels 0..depth-1 — the elbow's own level is not one of
                // them (its `tee` says whether it continues).
                pass_through: (0..depth.saturating_sub(1))
                    .filter(|level| seen[level + 1])
                    .collect(),
            };
        }
        seen[depth] = true;
        for level in seen.iter_mut().skip(depth + 1) {
            *level = false;
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn guides(depths: &[usize]) -> Vec<(Option<usize>, bool, Vec<usize>)> {
        guides_for(depths)
            .into_iter()
            .map(|guide| (guide.elbow_at, guide.tee, guide.pass_through))
            .collect()
    }

    /// A flat list draws nothing.
    #[test]
    fn roots_draw_no_connector() {
        assert_eq!(
            guides(&[0, 0, 0]),
            vec![(None, false, vec![]); 3]
        );
        assert_eq!(guides(&[]), Vec::new());
    }

    /// A straight chain: every child is its parent's LAST child, so every
    /// connector is an elbow and nothing passes through.
    #[test]
    fn a_chain_is_all_elbows() {
        assert_eq!(
            guides(&[0, 1, 2, 3]),
            vec![
                (None, false, vec![]),
                (Some(0), false, vec![]),
                (Some(1), false, vec![]),
                (Some(2), false, vec![]),
            ]
        );
    }

    /// Siblings under one parent: every one but the LAST is a tee.
    #[test]
    fn siblings_tee_until_the_last_one() {
        assert_eq!(
            guides(&[0, 1, 1, 1]),
            vec![
                (None, false, vec![]),
                (Some(0), true, vec![]),
                (Some(0), true, vec![]),
                (Some(0), false, vec![]),
            ]
        );
    }

    /// A grandchild under a NON-last child passes its grandparent's gutter
    /// through; under the last child it does not.
    #[test]
    fn a_grandchild_passes_a_continuing_ancestor_through() {
        // root ├ a │ └ a1 └ b └ b1
        assert_eq!(
            guides(&[0, 1, 2, 1, 2]),
            vec![
                (None, false, vec![]),
                (Some(0), true, vec![]),
                // `a` still has the sibling `b` below → level 0 passes through.
                (Some(1), false, vec![0]),
                (Some(0), false, vec![]),
                // `b` is the last child → nothing passes through.
                (Some(1), false, vec![]),
            ]
        );
    }

    /// Three levels deep: only the ancestors that CONTINUE draw.
    #[test]
    fn only_continuing_ancestors_pass_through() {
        // r ├ a │ ├ a1 │ │ └ x │ └ a2 └ b
        let depths = [0, 1, 2, 3, 2, 1];
        assert_eq!(
            guides(&depths),
            vec![
                (None, false, vec![]),
                (Some(0), true, vec![]),
                (Some(1), true, vec![0]),
                // `x` sits under `a1` (which has `a2` below) under `a`
                // (which has `b` below): both ancestors pass through.
                (Some(2), false, vec![0, 1]),
                (Some(1), false, vec![0]),
                (Some(0), false, vec![]),
            ]
        );
    }

    /// A row's depth rides its connector.
    #[test]
    fn the_depth_comes_off_the_elbow() {
        let depths: Vec<usize> = guides_for(&[0, 1, 2, 1])
            .iter()
            .map(Guides::depth)
            .collect();
        assert_eq!(depths, vec![0, 1, 2, 1]);
    }

    /// A row whose depth JUMPS back to a root closes every deeper level.
    #[test]
    fn a_new_root_closes_the_previous_subtree() {
        assert_eq!(
            guides(&[0, 1, 2, 0, 1]),
            vec![
                (None, false, vec![]),
                (Some(0), false, vec![]),
                (Some(1), false, vec![]),
                (None, false, vec![]),
                (Some(0), false, vec![]),
            ]
        );
    }
}
